src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5028,6 +5028,56 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
 EXTERN_C void JIT_PInvokeBegin(InlinedCallFrame* pFrame);
 EXTERN_C void JIT_PInvokeEnd(InlinedCallFrame* pFrame);
 
+// Forward declaration
+EXTERN_C void STDCALL ReversePInvokeBadTransition();
+
+// This is a slower version of the reverse PInvoke enter function.
+NOINLINE static void JIT_ReversePInvokeEnterRare(FastReversePInvokeFrame* frame)
+{
+    _ASSERTE(frame != NULL);
+
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
+
+    // Verify the current thread isn't in COOP mode.
+    if (thread->PreemptiveGCDisabled())
+        ReversePInvokeBadTransition();
+
+    thread->DisablePreemptiveGC();
+    frame->currentThread = thread;
+}
+
+EXTERN_C void JIT_ReversePInvokeEnter(FastReversePInvokeFrame* frame)
+{
+    _ASSERTE(frame != NULL);
+    Thread* thread = GetThreadNULLOk();
+
+    // If a thread instance exists and is in the
+    // correct GC mode attempt a quick transition.
+    if (thread != NULL
+        && !thread->PreemptiveGCDisabled())
+    {
+        // Manually inline the fast path in Thread::DisablePreemptiveGC().
+        thread->m_fPreemptiveGCDisabled.StoreWithoutBarrier(1);
+        if (g_TrapReturningThreads.LoadWithoutBarrier() == 0)
+        {
+            frame->currentThread = thread;
+            return;
+        }
+    }
+
+    JIT_ReversePInvokeEnterRare(frame);
+}
+
+EXTERN_C void JIT_ReversePInvokeExit(FastReversePInvokeFrame* frame)
+{
+    _ASSERTE(frame != NULL);
+    _ASSERTE(frame->currentThread == GetThread());
+
+    frame->currentThread->EnablePreemptiveGC();
```

Comment:
It should be sufficient to just clear `m_fPreemptiveGCDisabled` like what the current delegate assembly helpers are doing. The extra `CatchAtSafePoint` should not be required.

Reply:
@jkotas I get the reason for the `*Enter()` case being unwrapped because of the fallback (i.e. `*Rare()`) that will handle this but the `*Exit()` doesn't have that. What is the reason this isn't needed? Personally, accessing fields in this way when public functions exist is a code smell and calls into question why these functions exist if we don't use them consistently. Perhaps this warrants a new function `EnablePreemptiveGCFast()`? I accept that your statement is correct to just set the bit, but why?

Reply:
> What is the reason this isn't needed? 

For the unmanaged/managed transitions, the effect of CatchAtSafePoint would be that the GC suspension will finish a bit faster in some cases. It is a trade-off 

It is a trade-off: Spend extra cycles on each transition and make GC suspension faster; or make each transition faster and make GC suspension a bit slower occasionally.

We are opting for the faster unmanaged/managed transitions everywhere else today, so this is for consistency. Also, the existing unmanaged/managed transitions are in assembly helpers, so there was no need to have a C++ method to do that.

> EnablePreemptiveGCFast

I do not think that having `EnablePreemptiveGCFast` method would be an improvement over just setting the field directly.

If you would like to make this pretty, the way to do that would be to encapsulate the transition in FORCEINLINE methods, similar to how it is done in CoreRT: https://github.com/dotnet/corert/blob/41557532f0ff3e4bb0393a5451cae65a041ad5ae/src/Native/Runtime/thread.h#L259